File: tests/networking/TestNetworking.pas
//
// TestNetworking.pas
//
// This unit contains unittests for the Networking unit.
//
// Copyright (c) 2003-2022 Matias Vara <matiasevara@gmail.com>
// All Rights Reserved
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//

program TestNetworking;

{$IFDEF FPC}
 {$mode delphi}
{$ENDIF}

uses
  Kernel,
  Process,
  Memory,
  Debug,
  Arch,
  Filesystem,
  Network,
  Console,
  VirtIO,
  VirtIOConsole,
  VirtIOVSocket;

const
  SERVICE_TIMEOUT = 1000;
  INIT_MEM = 2048;

var
  total, i, total_t, new_size, curr_size: LongInt;
  netdriver, tmp, mode: PChar;
  HttpServer, HttpClient: PSocket;
  buf: array[0..255] of Byte;
  buff, buff_i: PChar;

// This test sends an array buf and expects to get the same array
// If not, it returns False
function TestClient: Boolean;
begin
  Result := False;
  HttpClient := SysSocket(SOCKET_STREAM);
  // Host CID
  HttpClient.DestIp := 2;
  HttpClient.DestPort := 80;
  HttpClient.Blocking := True;
  if SysSocketConnect(HttpClient) then
  begin
    // generate a sequence of numbers
    for i := 0 to sizeof(buf)-1 do
      buf[i] := i;
    SysSocketSend(HttpClient, @buf[0], sizeof(buf), 0);
    while true do
    begin
      if not SysSocketSelect(HttpClient, SERVICE_TIMEOUT) then
        Break;
      total := SysSocketRecv(HttpClient, @buf[0], sizeof(buf), 0);
    end;
    SysSocketClose(HttpClient);
    // check if the sequence is correctly received
    for i:= 0 to sizeof(buf)-1 do
      if buf[i] <> i Then
        Exit;
    Result := True;
  end;
end;

begin
  netdriver := GetKernelParam(0);
  mode := GetKernelParam(1);
  DedicateNetworkSocket(netdriver);
  if StrCmp(mode, 'client', StrLen('client')) then
  begin
    If TestClient = False then
      WriteDebug('TestClient-%d: FAILED\n', [0])
    else
      WriteDebug('TestClient-%d: PASSED\n', [0]);
    Exit
  end else if StrCmp(mode, 'server', StrLen('server')) then
  begin
    HttpServer := SysSocket(SOCKET_STREAM);
    HttpServer.Sourceport := 80;
    HttpServer.Blocking := True;
    SysSocketListen(HttpServer, 50);
    // server mode
    while true do
    begin
      curr_size := INIT_MEM;
      buff := ToroGetMem(curr_size);
      buff_i := buff;
      total_t := 0;
      HttpClient := SysSocketAccept(HttpServer);
      while true do
      begin
        if not SysSocketSelect(HttpClient, SERVICE_TIMEOUT) then
          Break;
        total := SysSocketRecv(HttpClient, @buf, sizeof(buf), 0);
        if total = 0 then
          Continue;
        if total_t + total > curr_size then
        begin
          Inc(new_size, curr_size + sizeof(buf));
          tmp := ToroGetMem(new_size);
          Move(buff_i^, tmp^, total_t);
          ToroFreeMem(buff_i);
          buff_i := tmp;
          buff := buff_i + total_t;
          curr_size := new_size;
        end;
        Inc(total_t, total);
        Move(buf, buff^, total);
        Inc(buff, total);
      end;
      SysSocketSend(HttpClient, buff_i, total_t, 0);
      SysSocketClose(HttpClient);
      ToroFreeMem(buff_i);
    end;
  end;
end.

File: tests/networking/testnetworking.py
#!/usr/bin/env python3
#
# testnetworking.py
#
# This script tests the networking stack by relying on two modes: as a client
# and as a server.
#
# Copyright (c) 2003-2022 Matias Vara <matiasevara@gmail.com>
# All Rights Reserved
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
import socket, random, sys

# send up to 1MB of chunk
REQ_MAX = 1024*1024

def generate_msg(size):
    msg = bytes(''.join((random.choice('abcdxyzpqr') for i in range(size))), 'UTF-8')
    return msg

def send_msg(cdi, port, msg, size):
    s = socket.socket(socket.AF_VSOCK, socket.SOCK_STREAM)
    response = bytearray()
    try:
        s.connect((cdi, port))
        print(size)
        s.sendall(msg)
        while len(response) < size:
            packet = s.recv(size - len(response))
            if not packet:
                break
            response.extend(packet)
    except socket.error:
        print("Some error has happened")
    finally:
        s.close()
    return response

# test server mode by sending chunks and waiting to get the same
def test_server(cid, port, req_max):
    req_len = 2
    while req_len < req_max:
        msg = generate_msg(req_len)
        response = send_msg(cid, port, msg, req_len)
        assert msg == response, 'failed at req_len=' + str(req_len) + '-' + str(len(msg)) + '-' + str(len(response))
        req_len *= 2
    return 0

# test client mode by receiving a 256 bytes long chunk and sending it back
def test_client(port):
    CID = socket.VMADDR_CID_HOST
    s = socket.socket(socket.AF_VSOCK, socket.SOCK_STREAM)
    s.bind((CID, port))
    s.listen()
    (conn, (remote_cid, remote_port)) = s.accept()
    while True:
        # TODO: make chunk size dinamic
        buf = conn.recv(256)
        if not buf:
            break
        conn.sendall(buf)
    conn.close()

if len(sys.argv) < 4:
    print("usage: testnetwork.py [cid] [port] [server|client]")
    exit(1)

cid = int(sys.argv[1])
port = int(sys.argv[2])
if sys.argv[3] == 'server':
    print("Launching server test ...")
    test_server(cid, port, REQ_MAX)
elif sys.argv[3] == 'client':
    print("Launching client test ...")
    test_client(int(port));
exit(0)
